lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ExecutorServiceCleanupBinder.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.base;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Binder;
+import com.google.inject.BindingAnnotation;
+import com.google.inject.Inject;
+import com.google.inject.Key;
+import com.google.inject.Scopes;
+import com.google.inject.multibindings.Multibinder;
+import jakarta.annotation.PreDestroy;
+
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+import static com.google.inject.multibindings.Multibinder.newSetBinder;
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+public class ExecutorServiceCleanupBinder
+{
+    public static ExecutorServiceCleanupBinder executorServiceCleanupBinder(Binder binder)
+    {
+        return new ExecutorServiceCleanupBinder(binder);
+    }
+
+    private final Multibinder<ExecutorService> multibinder;
+
+    public ExecutorServiceCleanupBinder(Binder binder)
+    {
+        multibinder = newSetBinder(binder, ExecutorService.class, ForCleanup.class);
+        binder.bind(Cleanup.class).asEagerSingleton();
```

Comment:
I'm not a guice expert, but I'd check whether binding as eager singleton means that it will instantiate with the multibinder in the state that it is at this point in time, or whether it will instantiate once it iterated all the modules for any additional multibinding.

Reply:
Since the cleanup class constructor depends on the set of executors, it will be created after this set (this is dependency in the graph)